### DISCUSSION TITLE
Incorrect documentation for authenticate_or_request_with_http_token

### DISCUSSION THREAD

**Discussion by collimarco** [rails/rails#51919] [Docs]
Ruby on Rails API documentation is incorrect for [`authenticate_or_request_with_http_token `](https://github.com/rails/rails/blob/747a03ba7722b6f0a7ce42e86cea83cf07a2e6ef/actionpack/lib/action_controller/metal/http_authentication.rb#L428):

```
# Authenticate using an HTTP Bearer token, or otherwise render an HTTP
# header requesting the client to send a Bearer token. For the authentication
# to be considered successful, +login_procedure+ should return a **non-nil**
# value. Typically, the authenticated user is returned.
```

The word `non-nil` is not the correct word to use here.

The correct word is `non-falsy` value, because **the login procedure fails if you return `nil` or `false`**.

Alternatively, the word `truthy` can also be used.

**Comment by zzak** [comment]
Just fyi,

https://edgeguides.rubyonrails.org/api_documentation_guidelines.html#booleans

> Please avoid terms like "truthy". Ruby defines what is true and false in the language, and thus those words have a technical meaning and need no substitutes.

**Reply by collimarco** [reply]
**Still the documentation is incorrect...**

**Based on the documentation above the method, if you return `false` (which is `non-nil`), login procedure will be considered successful.**

**But THAT IS NOT WHAT HAPPENS!**

If you return `false`, the login procedure will be considered UNSUCCESSFUL.

You need to make that clear in the documentation.

NOTE: I am suggesting to fix the documentation, not the method implementation, which I think it does the correct thing.

Is now clearer?

**Reply by collimarco** [reply]
Also note that "falsy" is not the same thing as "false".

"Falsy" is a term used to indicate both "false" and "nil" at the same time.

**Reply by zzak** [reply]
Sorry, I wasn't trying to say you're wrong just that there is a precedent for not using "truthy" or "falsy" in the documentation.

If you can find a way to convey that please open a PR, I haven't looked into this beyond that yet.

**Reply by collimarco** [reply]
Personally I would use the word "non-falsy" in this context. If you say "non-false", the meaning is different...

However if you don't want to use the word "falsy"/"truthy", we can rephrase in this way:

> For the authentication to be considered successful, +login_procedure+ should return a non-nil value.
> If you return nil or false, then the authentication will be considered unsuccessful.

@zzak Can you please reopen the related issue so that we can fix this?

**Reply by collimarco** [reply]
@zzak As requested, I also opened the pull request here: https://github.com/rails/rails/pull/51927

**Comment by JuanVqz** [comment]
This discussion ended up merging a PR, should it be closed?
